### ISSUE TITLE
Svelte imports autocompletion suggestion

### ISSUE THREAD

**Issue by 0xDjole** [ms-jpq/coq_nvim#161]
Suggestions for import paths in Svelte files are incorrect.

I think it takes the reference of current working directory opened in Vim and puts it as reference for all nested files.

For example:
c ( some other directory not in scope )
a ( present working directory )
   b
       file1 ( import something from '../' will open suggestions for c folder because it takes a as the point of reference  )

I hope I am correct about my thinking behind the error.
From what I've seen it doesn't happen in regular typescript files.

**Comment by ms-jpq** [comment]
can you please show me an example so I can reproduce.

the `path` source order is actually:

```
# this early returns

-> try relative to <base>
---> try expand `~`
---> try expand `$VARIABLE` and `${VARIABLE}`
```

where `<base>` is "parent of current file", and "cwd", they are looked up in parallel.

and then everything gets merged together.

so what could have happened was that `coq` finds too many files, and the files you want dont get shown.

can you please try manual completion (default `<c-space>`), and see if it finds the thing that you are looking for?

---

anyways, `path` source actually sucks, since right now it doesn't handle two files on the same line, so I will work on that.

**Comment by 0xDjole** [comment]
See Counter.svelte. It is nested deep.
Completions I get are from parent directory from my svelte-boilerplate directory, not parent of Counter.svelte.
My working directory is svelte-boilerplate. It takes that as base....Literally returning incorrect suggestions.

![2021-08-30_15-59](https://user-images.githubusercontent.com/36204260/131351177-110aff8b-e77d-4c7d-b119-bc627129a75a.png)

**Comment by 0xDjole** [comment]
I tried manual completion, returns same thing :D

**Comment by dnaaun** [comment]
I've no idea how Svelte works, but in Python for example, when you import from a file that is in the same directory as another file, you do `from .otherFile import some_function` instead of `from ..otherFile import some_function`.

Assuming that in Svelte, you would use `import something from '../otherFile`, I want to note that, currently, you're getting suggestions from the "paths" source, and not the "lsp" source. The LSP should be aware of how Svelte imports work and make proper suggestions. The "paths" source literally just lists all the files under "../", which by unix convention means "one directory up".

Could you install [a Svelte LSP](https://github.com/neovim/nvim-lspconfig/blob/master/CONFIG.md#svelte), and then, if necessary play with the `coq_settings.clients.paths.weight_adjust` and `coq_settings.clients.lsp.weight_adjust` settings to make sure that LSP suggestions show up with higher priority, and then report the results?

**Comment by 0xDjole** [comment]
I haven't found coq_settings.clients.paths.weight_adjust and coq_settings.clients.lsp.weight_adjust settings, they are not documented so I am not sure what to put.

Please see what I am doing in the screenshot. I am going up one directory in the import. Dropdown shows. It displays `../anarchy` for example as first result. THAT IS INCORRECT. `anarchy` folder is one level above `svelte-boilerplate` folder. 

I am importing from `Counter.svelte` file. The parent of that file is folder `lib`.  If I go one step out of that `lib` directory I should see results from `src` directory. Everything NOT in `src` directory is INCORRECT

Now that that is hopefully clear, wanted suggestions should be all files from `src` folder.

It's not even question about weights and priority tho it would help. Point is that results are incorrect. If you were to put suggested imports, it wouldn't even work.

Issue could be from svelte-lsp itself I suppose. I don't know how any of those things work in the background...

**Comment by ms-jpq** [comment]
Oh, so is this like an issue where `paths` source is returning extra results you don't want?

I can add an option to say `only return results relative to cwd` and `only return results relative to current file`.

I will keep the defaults identical because I find a lot of the time, both are really helpful

---

like for you, you only want things relative to current file and not cwd right?

**Comment by 0xDjole** [comment]
Correct. Do that. Only correct imports are the relative ones and the global ones coming from node_modules. `CWD` is incorrect. 

I just checked in plain typescript again, and realized it is the same issue. So in both svelte and regular TS should be fixed. And actually it should be fixed on all js files. ( jsx, tsx, js, ts, .svelte )

Keep in mind global imports ( the one without  `.` or `..` ) should be watching `node_modules` directory. So for global imports, `cwd` should be `node_modules`. To make sure it is correct. That would happen only if `package.json` is present.

So global imports should look in `node_modules` if `package.json` and `node_modules` exist.

It's not what I want, it's whats correct.

**Comment by ms-jpq** [comment]
oh I think you have it confused,

`PATHS` source doesn't care about file types at all, what you want is like @davidatbu  said, is the `LSP`'s job.

`PATHS` source is purely a dumb "look at my file system layout" type of deal, it has zero semantic awareness of the document you are editing aside from that.

if you don't like it, you can disable `PATHS` entirely, and let `LSP` handle the import resolutions.

**Comment by 0xDjole** [comment]
How can I disable it only for svelte? I only have this:

`lsp.svelte.setup(coq.lsp_ensure_capabilities({ on_attach = on_attach }))`

**Comment by dnaaun** [comment]
> Everything NOT in src directory is INCORRECT

Ah yes. My bad. I misread your original comment.

> I haven't found coq_settings.clients.paths.weight_adjust and coq_settings.clients.lsp.weight_adjust settings

Typing in `:COQhelp sources` will pull up the help.

> weights and priority ... would help.

Here is how you'd set priority/weights for various completion sources(note that this is in lua).

```lua
vim.g.coq_settings = {
  clients = {
    lsp = {
      enabled = false,
      weight_adjust = 2,
    },
    buffers = {
      enabled = true,
      weight_adjust = 0,
    },
    paths = {
      enabled = true,
      weight_adjust = -1
    }
  }
}
```

> Issue could be from svelte-lsp itself I suppose. I don't know how any of those things work in the background...

This issue is definitely not from svelte-lsp since coq is not getting the completion results shown in the screenshot from the LSP (if it were, it would say `LSP` instead of `PATH` for the completion items from the LSP in the popup menu).

Like @ms-jpq said, I think you should try to get LSP completions. 

> How can I disable [the paths source] only for svelte?

@ms-jpq can correct me on this, but there's currently no way to do a per-filetype setting for which sources are enabled.

My suggestion for you is to play around with the `weight_adjust` setting of the various completion sources as I've shown you above to try to get LSP suggestions get ranked way higher than path suggestions. I find that it works quite well whenever an LSP is present, it's suggestions take priority (which is what I usually want), and when an LSP is not present, I get the "dumb" path suggestions as usual.

**Comment by ms-jpq** [comment]
now that https://github.com/ms-jpq/coq_nvim/pull/170 is merged.

@0xDjole just set `coq_settings.clients.paths.resolution` to `["file"]`, and you are good.

@davidatbu added an option so OP doesn't need to fiddle around with weights.

**Comment by ms-jpq** [comment]
thanks ! :D

**Comment by 0xDjole** [comment]
Now it works. It's perfect.

Ideally what you want is to have this a default behavior for certain files types...Not sure how much work would require.

But I can tell you 100% not a single js/ts dev in the world would like these extra import suggestions that are incorrect...
